pkg/util/eventutil/events.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+// EventReporter is custom events reporter type which allows user to limit the events
+type EventReporter struct {
+	mux            sync.Mutex
+	recorder       record.EventRecorder
+	reportedEvents map[string]*eventObject
+
+	// report events x times where x is count
+	count int
+
+	// report events after x minutes
+	eventReportAfterMinutes int
+}
+
+// NewEventReporter returns EventReporter object
+func NewEventReporter(recorder record.EventRecorder, maxCountInGivenTime, reportAfter int) *EventReporter {
+	er := &EventReporter{
+		recorder:                recorder,
+		count:                   maxCountInGivenTime,
+		eventReportAfterMinutes: reportAfter,
+	}
+
+	er.reportedEvents = map[string]*eventObject{}
+
+	return er
+}
+
+// Report records a events if eventReportAfterMinutes has passed or events occurred less than count
+func (rep *EventReporter) Report(instance runtime.Object, eventType, eventReason, msg string) {
+	rep.mux.Lock()
+	defer rep.mux.Unlock()
+
+	objMeta, err := meta.Accessor(instance)
+
+	if err != nil {
+		return
+	}
+
+	eventKey := fmt.Sprintf("%s:%s:%s:%s", objMeta.GetName(), eventType, eventReason, msg)
+
+	eventobj, ok := rep.reportedEvents[eventKey]
+	if !ok {
+		eventobj = &eventObject{eventsReportedAt: time.Now(), eventCount: 1}
+		rep.reportedEvents[eventKey] = eventobj
+		rep.recorder.Event(instance, eventType, eventReason, msg)
+	} else if eventobj.eventCount < rep.count {
+		eventobj.eventCount++
+		rep.recorder.Event(instance, eventType, eventReason, msg)
+	} else if eventobj.eventsReportedAt.Add(time.Minute * time.Duration(rep.eventReportAfterMinutes)).Before(time.Now()) {
+		eventobj.eventCount = 1
+		eventobj.eventsReportedAt = time.Now()
+		rep.recorder.Event(instance, eventType, eventReason, msg)
+	}
```

Comment:
Can you show the output of an example event that you see from `kubectl describe` after testing this? Also, have you been able to test these different cases manually?